tutorials/dataframe/df107_SingleTopAnalysis.py

```diff
@@ -0,0 +1,245 @@
+    df[s] = df[s].Filter("trigE || trigM")\
+                 .Filter("met_et > 30000")
+
+    # Perform preselection of highly isolated leptons
+    df[s] = df[s].Define("goodlep", "lep_isTightID && lep_pt > 35000 && lep_ptcone30 / lep_pt < 0.1 && lep_etcone20 / lep_pt < 0.1")\
+                 .Filter("Sum(goodlep) > 0")
+
+    # Find a single good lepton, otherwise return -1 as index
+    df[s] = df[s].Define("idx_lep", "FindGoodLepton(goodlep, lep_type, lep_pt, lep_eta, lep_phi, lep_E, lep_trackd0pvunbiased, lep_tracksigd0pvunbiased, lep_z0)")\
+                 .Filter("idx_lep != -1")
+
+    # Compute transverse mass of the W boson using the missing transverse energy and the good lepton
+    # Use only events with a transverse mass of the reconstruced W boson larger than 60 GeV
+    df[s] = df[s].Define("mtw", "ComputeTransverseWMass(lep_pt[idx_lep], lep_eta[idx_lep], lep_phi[idx_lep], lep_E[idx_lep], met_et, met_phi)")\
+                 .Filter("mtw > 60000")
+
+    # Perform preselection of jets
+    df[s] = df[s].Filter("Sum(jet_pt > 30000 && abs(jet_eta) < 2.5) > 0")
+
+    # Select events with two good jets and one b-jet and find the indices in the collections
+    df[s] = df[s].Define("goodjet", "jet_pt > 60000 || abs(jet_eta) > 2.4 || jet_jvt > 0.59")\
+                 .Filter("Sum(goodjet) == 2")\
+                 .Define("goodbjet", "goodjet && jet_MV2c10 > 0.8244273")\
+                 .Filter("Sum(goodbjet) == 1")\
+                 .Define("idx_tagged", "ArgMax(goodjet && goodbjet)")\
+                 .Define("idx_untagged", "ArgMax(goodjet && !goodbjet)")
+
+    # Select events based on the jet kinematics and the scalar sum of the transverse momentum
+    # from the lepton, jets and met above 195 GeV
+    df[s] = df[s].Filter("abs(jet_eta[idx_untagged]) > 1.5 && abs(jet_eta[idx_tagged] - jet_eta[idx_untagged]) > 1.5")\
+                 .Filter("lep_pt[idx_lep] + jet_pt[idx_tagged] + jet_pt[idx_untagged] + met_et > 195000")
+
+# Compute luminosity, scale factors and MC weights for simulated events
+lumi = 10064.0
+for s in samples:
+    if "data" in s:
+        df[s] = df[s].Define("weight", "1.0")
+    else:
+        # The single top MC weights are either 1 or -1
+        if "single" in s: stop_norm = "mcWeight / abs(mcWeight)"
+        else: stop_norm = "mcWeight"
+        df[s] = df[s].Define("weight", "scaleFactor_ELE * scaleFactor_MUON * scaleFactor_LepTRIGGER * scaleFactor_PILEUP * scaleFactor_BTAG * {} * {} / {} * {}".format(stop_norm, xsecs[s], sumws[s], lumi))
+
+# Reconstruct the top mass from the lepton, the missing transverse energy and the b-jet
+# Restrict the top_mass to events with the visible component above 150 GeV and finally book
+# a histogram.
+ROOT.gInterpreter.Declare("""
+float ComputeTopMass(float lep_pt, float lep_eta, float lep_phi, float lep_e, float jet_pt, float jet_eta, float jet_phi, float jet_e, float met_et, float met_phi)
```

Comment:
`ComputeVisibleTopMass`?

Reply:
It's not the visible top mass! We use the MET to approximate the neutrino, which gives the full top mass. It's a little bit cheated because we miss the z component of the neutrino, but well ...